.github/workflows/docker-publish.yml

```diff
@@ -1,95 +1,57 @@
+    tags:
+      - 'v[0-9]+.[0-9]+.[0-9]+'
+      - '[0-9]+.[0-9]+.[0-9]+'
+      - '[0-9]+.[0-9]+.[0-9]+-*'
   pull_request:
-    branches: [ "master", "develop" ]
+    branches: [ "main", "master", "develop" ]
+  release:
+    types: [published]
+
 env:
-  # Use docker.io for Docker Hub if empty
   REGISTRY: ghcr.io
-  # github.repository as <account>/<repo>
   IMAGE_NAME: ${{ github.repository }}
 
-
 jobs:
   build:
-
     runs-on: ubuntu-latest
     permissions:
       contents: read
       packages: write
-      # This is used to complete the identity challenge
-      # with sigstore/fulcio when running outside of PRs.
-      id-token: write
 
     steps:
-      - name: Checkout repository
-        uses: actions/checkout@v4
-
-      # Install the cosign tool except on PR
-      # https://github.com/sigstore/cosign-installer
-      - name: Install cosign
-        if: github.event_name != 'pull_request'
-        uses: sigstore/cosign-installer@59acb6260d9c0ba8f4a2f9d9b48431a222b68e20 #v3.5.0
-        with:
-          cosign-release: 'v2.2.4'
-
-      # Set up BuildKit Docker container builder to be able to build
-      # multi-platform images and export cache
-      # https://github.com/docker/setup-buildx-action
-      - name: Set up Docker Buildx
-        uses: docker/setup-buildx-action@f95db51fddba0c2d1ec667646a06c2ce06100226 # v3.0.0
-
-      # Login against a Docker registry except on PR
-      # https://github.com/docker/login-action
-      - name: Log into registry ${{ env.REGISTRY }}
-        if: github.event_name != 'pull_request'
-        uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3.0.0
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v3
         with:
           registry: ${{ env.REGISTRY }}
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
+    
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
 
-      # Extract metadata (tags, labels) for Docker
-      # https://github.com/docker/metadata-action
       - name: Extract Docker metadata
         id: meta
-        uses: docker/metadata-action@96383f45573cb7f253c731d3b3ab81c87ef81934 # v5.0.0
+        uses: docker/metadata-action@v5
         with:
           images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
 
-      # Build and push Docker image with Buildx (don't push on PR)
-      # https://github.com/docker/build-push-action
       - name: Build and push Docker image
         id: build-and-push
-        uses: docker/build-push-action@0565240e2d4ab88bba5387d719585280857ece09 # v5.0.0
+        uses: docker/build-push-action@v5
         with:
           context: .
-          push: ${{ github.event_name != 'pull_request' }}
+          push: true
           tags: ${{ steps.meta.outputs.tags }}
           labels: ${{ steps.meta.outputs.labels }}
```

Comment:
What tags are the images going to get for merges, PRs, and releases? It's not clear to me from this

Reply:
@MrCreosote Check out https://github.com/bio-boris/BuildMeARiver/pkgs/container/buildmeariver to see what they would look like

Reply:
Is not clear to me the difference between pr-n, latest, and main. And n will just automatically increment from a head branch?

Reply:
* PRs are builds from a pull request that has not merged in yet.
* It looks like latest points to the latest release
* main points to the main branch
* develop points to the develop branch

@Xiangs18

Reply:
Not sure why we need `tags` and `labels` in this step. I thought image name is declared in the previous step?

Reply:
The image name is extracted in the previous step, and applied in the next step. 

To read more about labels see 
* https://docs.docker.com/config/labels-custom-metadata/
* https://docs.docker.com/build/ci/github-actions/manage-tags-labels/

Reply:
```
docker inspect ghcr.io/jgi-kbase/idmappingservice:pr-132 | jq  -r '.[0].Config.Labels'
{
  "maintainer": "Steve Chan sychan@lbl.gov",
  "org.label-schema.build-date": "",
  "org.label-schema.schema-version": "1.0.0-rc1",
  "org.label-schema.vcs-ref": "",
  "org.label-schema.vcs-url": "https://github.com/jgikbase/IDMappingService.git",
  "org.opencontainers.image.created": "2024-06-24T19:33:44.496Z",
  "org.opencontainers.image.description": "Service for mapping data IDs between namespaces.",
  "org.opencontainers.image.licenses": "MIT",
  "org.opencontainers.image.revision": "e7720c8e5648848d088fd18bb08239812dd3d756",
  "org.opencontainers.image.source": "https://github.com/jgi-kbase/IDMappingService",
  "org.opencontainers.image.title": "IDMappingService",
  "org.opencontainers.image.url": "https://github.com/jgi-kbase/IDMappingService",
  "org.opencontainers.image.version": "pr-132",
  "us.kbase.vcs-branch": "develop"
}
```
Here are the labels for this pr

Reply:
is `Extract Docker metadata` step optional?

Reply:
No, it is required in order to get the metadata without having to implement our own logic for getting the metadata.

Reply:
It makes sense now. Thank you!!